include/private/net/libssl.hpp

```diff
@@ -0,0 +1,327 @@
+        Context instance on success.
+    */
+    template <MK_MOCK(SSL_CTX_new), MK_MOCK(SSL_CTX_load_verify_locations)
+#if (defined LIBRESSL_VERSION_NUMBER && LIBRESSL_VERSION_NUMBER >= 0x2010400fL)
+                                          ,
+              MK_MOCK(SSL_CTX_load_verify_mem)
+#endif
+              >
+    static ErrorOr<Var<Context>> make(std::string ca_bundle_path,
+                                      Var<Logger> logger) {
+        // Implementation note: we need to get the library singleton as early
+        // as here to make sure OpenSSL internals are initialized.
+        auto library = Library::global();
+        logger->debug("ssl: creating SSL_CTX with bundle: '%s'",
+                      ca_bundle_path.c_str());
+        SSL_CTX *ctx = SSL_CTX_new(SSLv23_client_method());
+        if (ctx == nullptr) {
+            logger->warn("ssl: failed to create SSL_CTX");
+            return SslCtxNewError();
+        }
+        /*
+         * Implementation note:
+         *
+         * We use SSLv23_client_method() because it is the most general method
+         * that applications SHOULD use according to the manual[1] and we
+         * disable by default deprecated protocols. We MAY want to enable them
+         * in the OONI's web_connectivity test so to reach more websites.
+         *
+         * [1] https://wiki.openssl.org/index.php/Manual:SSL_CTX_new(3)
+         */
+        SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+        if (ca_bundle_path != "") {
+            if (!SSL_CTX_load_verify_locations(ctx, ca_bundle_path.c_str(),
+                                               nullptr)) {
+                logger->warn("ssl: failed to load verify location");
+                SSL_CTX_free(ctx);
+                return SslCtxLoadVerifyLocationsError();
+            }
+        } else {
+#if (defined LIBRESSL_VERSION_NUMBER && LIBRESSL_VERSION_NUMBER >= 0x2010400fL)
+            std::vector<uint8_t> bundle = builtin_ca_bundle();
+            logger->debug("ssl: using builtin libressl's ca bundle");
+            if (!SSL_CTX_load_verify_mem(ctx, bundle.data(), bundle.size())) {
+                logger->warn("ssl: failed to load default ca bundle");
+                SSL_CTX_free(ctx);
+                return SslCtxLoadVerifyMemError();
+            }
+#else
+            SSL_CTX_free(ctx);
+            return MissingCaBundlePathError();
+#endif
+        }
+        SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, nullptr);
+        Var<Context> context{new Context};
+        context->ctx_ = ctx;
+        return context;
+    }
+
+  private:
+    Context() {}
+    SSL_CTX *ctx_ = nullptr;
+};
+
+/*!
+    \brief Thread-safe cache for constructing `SSL *` objects.
+
+    \remark Unless you need low-level control, you SHOULD really use
+    this API for creating an `SSL *` when connecting to an host.
+*/
+template <size_t max_cache_size = 64> class Cache {
```

Comment:
This is not refactored, but rather newly written code functionally equivalent to `make_ssl_ctx()` but more clear.